caddy/app.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2024 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package caddy provides an app and handler for Caddy Server (https://caddyserver.com/)
+// allowing it to turn any handler into one supporting the Vulcain protocol.
+
+package caddy
+
+import (
+	"errors"
+	"log/slog"
+
+	outline_prometheus "github.com/Jigsaw-Code/outline-ss-server/prometheus"
+	outline "github.com/Jigsaw-Code/outline-ss-server/service"
+	"github.com/caddyserver/caddy/v2"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+func init() {
+	caddy.RegisterModule(OutlineApp{})
+}
+
+const outlineModuleName = "outline"
+
+type ShadowsocksConfig struct {
+	ReplayHistory int `json:"replay_history,omitempty"`
+}
+
+type OutlineApp struct {
+	ShadowsocksConfig *ShadowsocksConfig `json:"shadowsocks,omitempty"`
+
+	ReplayCache outline.ReplayCache
+	logger      *slog.Logger
+	Metrics     outline.ServiceMetrics
+	buildInfo   *prometheus.GaugeVec
+}
+
+var (
+	_ caddy.App         = (*OutlineApp)(nil)
+	_ caddy.Provisioner = (*OutlineApp)(nil)
+)
+
+func (OutlineApp) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID:  outlineModuleName,
+		New: func() caddy.Module { return new(OutlineApp) },
```

Comment:
Something seems wrong here. Is `OutlineApp` both an App and a Module?
You create an OutlineApp to then create another OutlineApp? That's a circular dependency and looks like a bug.

Reply:
This is so silly... I think we should create something like:

```go
type ModuleRegistration caddy.ModuleInfo

func (m *ModuleRegistration) CaddyModule() caddy.ModuleInfo {
  return m
}

func init() {
  caddy.RegisterModule(ModuleRegistration{
    ID: "outline"
    New: func() caddy.Module { return new(OutlineApp) }
  })
}
```

Reply:
> Something seems wrong here. Is `OutlineApp` both an App and a Module?

In Caddy apps are modules. So yes, `OutlineApp` is both an `App` and a `Module`.

> You create an OutlineApp to then create another OutlineApp? That's a circular dependency and looks like a bug.

Yeah it's a bit of an odd API. I don't think we can get around it though; `OutlineApp` must implement `caddy.Module`.

Reply:
Discussed offline. I misunderstood the suggestion. Will split them out.